internal/login/login.go

```diff
@@ -1,18 +1,193 @@
 package login
 
 import (
+	"context"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/ecdh"
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
 	"fmt"
+	"io"
+	"net/http"
 	"os"
+	"os/exec"
+	"strconv"
 	"strings"
+	"time"
 
+	"github.com/google/uuid"
 	"github.com/spf13/afero"
 	"github.com/supabase/cli/internal/utils"
 	"github.com/supabase/cli/internal/utils/credentials"
 )
 
-func Run(stdin *os.File, fsys afero.Fs) error {
-	accessToken := PromptAccessToken(stdin)
-	return utils.SaveAccessToken(accessToken, fsys)
+type RunParams struct {
+	Token       string
+	Name        string
+	OpenBrowser bool
+	Fsys        afero.Fs
+}
+
+type AccessTokenResponse struct {
+	SessionId   string `json:"id"`
+	CreatedAt   string `json:"created_at"`
+	AccessToken string `json:"access_token"`
+	PublicKey   string `json:"public_key"`
+	Nonce       string `json:"nonce"`
+}
+
+const defaultRetryAfterSeconds = 2
+const decryptionErrorMsg = "cannot decrypt access token"
+
+var loggedInMsg = "You are now logged in. " + utils.Aqua("Happy coding!")
+
+func decryptAccessToken(accessTokenResponse AccessTokenResponse, curve ecdh.Curve, privateKey *ecdh.PrivateKey) (string, error) {
+	decodedAccessToken, err := hex.DecodeString(accessTokenResponse.AccessToken)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	nonce, err := hex.DecodeString(accessTokenResponse.Nonce)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	publicKey, err := hex.DecodeString(accessTokenResponse.PublicKey)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	remotePublicKey, err := curve.NewPublicKey(publicKey)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	secret, err := privateKey.ECDH(remotePublicKey)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	block, err := aes.NewCipher(secret)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	decryptedAccessToken, err := aesgcm.Open(nil, nonce, decodedAccessToken, nil)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	return string(decryptedAccessToken), nil
+}
+
+func pollForAccessToken(url string) (AccessTokenResponse, error) {
```

Comment:
Let's add more unit tests for this. It would help with future refactoring if we change the http client.

You can check the uncovered lines here https://coveralls.io/builds/63814279/source?filename=internal%2Flogin%2Flogin.go#L90